wallet-new/src/Cardano/Wallet/Kernel/Wallets.hs

```diff
@@ -222,32 +240,45 @@ updatePassword pw hdRootId oldPassword newPassword = do
+             -- down to the 'Keystore' to ensure atomicity.
+             let pwdCheck = maybe False (const True) . checkPassMatches oldPassword
+
+             -- STEP 2: Compute the new key using the cryptographic primitives
              --         we have. This operation doesn't change any state, it
              --         just recomputes the new 'EncryptedSecretKey' in-place.
-             genNewKey <- changeEncPassphrase oldPassword newPassword esk
-             let mbNewKey = maybeToRight (UpdateWalletPasswordChangeFailed hdRootId) $
+             genNewKey <- changeEncPassphrase oldPassword newPassword oldKey
+             let mbNewKey = maybeToRight (UpdateWalletPasswordOldPasswordMismatch hdRootId) $
                             genNewKey
 
-             case pwdCheck >> mbNewKey of
+             case mbNewKey of
                   Left e -> return (Left e)
                   Right newKey -> do
-                      -- STEP 4: Update the keystore, atomically.
-                      Keystore.replace wId newKey keystore
-                      -- STEP 5: Update the timestamp in the wallet data storage.
-                      lastUpdateNow <- InDb <$> getCurrentTimestamp
-                      let hasSpendingPassword = HD.HasSpendingPassword lastUpdateNow
-                      res <- update' (pw ^. wallets)
-                                     (UpdateHdRootPassword hdRootId hasSpendingPassword)
-                      case res of
-                           Left e ->
-                               return $ Left (UpdateWalletPasswordUnknownHdRoot e)
-                           Right (db, hdRoot') -> return $ Right (db, hdRoot')
+                      -- STEP 3: Update the keystore, atomically.
+                      swapped <- Keystore.compareAndReplace wId pwdCheck newKey keystore
+                      case swapped of
+                           -- We failed, the password changed in the
+                           -- meantime, and the user needs to repeat the
+                           -- operation.
+                           Keystore.PredicateFailed ->
+                               return $ Left (UpdateWalletPasswordOldPasswordMismatch hdRootId)
+                           -- We failed, in the meantime the user deleted the
+                           -- key.
+                           Keystore.OldKeyLookupFailed -> do
+                               return $ Left (UpdateWalletPasswordKeystoreChangedInTheMeantime hdRootId)
+                           Keystore.Replaced -> do
+                               -- STEP 4: Update the timestamp in the wallet data storage.
+                               -- If we get interrupted here by an asynchronous exception the
+                               -- price we will pay would be a slightly incorrect notion of
+                               -- "how long ago we did change the password", but it won't
+                               -- compromise the integrity of the system.
```

Comment:
TODO: Add a comment here in the code to mention the subtle case of "NoPassword" and also mention we will revisit all those 3 functions with a better versions of `Keystore.insert/replace/delete` which would allow an `IO` action to run if the action on the keystore is successful.